src/plugins/kibana/server/lib/__tests__/create_mappings_from_pattern_fields.js

```diff
@@ -0,0 +1,81 @@
+const createMappingsFromPatternFields = require('../create_mappings_from_pattern_fields');
+const expect = require('expect.js');
+const _ = require('lodash');
+
+let testFields;
+
+describe('createMappingsFromPatternFields', function () {
+
+  beforeEach(function () {
+    testFields = [
+      {
+        'name': 'ip',
+        'type': 'ip'
+      },
+      {
+        'name': 'geo.coordinates',
+        'type': 'geo_point'
+      },
+      {
+        'name': 'agent',
+        'type': 'string'
+      },
+      {
+        'name': 'bytes',
+        'type': 'number'
+      }
+    ];
+  });
+
+  it('should throw an error if the argument is empty', function () {
+    expect(createMappingsFromPatternFields).to.throwException(/argument must not be empty/);
+  });
+
+  it('should not modify the original argument', function () {
+    const testFieldClone = _.cloneDeep(testFields);
+    const mappings = createMappingsFromPatternFields(testFields);
+
+    expect(mappings.ip).to.not.be(testFields[0]);
+    expect(_.isEqual(testFields, testFieldClone)).to.be.ok();
+  });
+
+  it('should set the same default mapping for all non-strings', function () {
+    let mappings = createMappingsFromPatternFields(testFields);
+
+    _.forEach(function (mapping) {
+      if (mapping.type !== 'string') {
+        expect(_.isEqual(mapping, {
```

Comment:
I'm using lodash's isEqual because I want to test for deep and strict equality. Unfortunately expect.js's is.eql() only checks for loose equality, and there's no assertion that's both recursive and strict.